toc.md

```diff
@@ -428,77 +429,234 @@ An additional backlog of 936 hours of image processing jobs everyday, along with
+
+In the next section we will look at how forking children processes impacts the consumption of the available memory.
 
 ##### Processes and Memory consumption
-We're able to create more processes, but surely that comes with a cost like everything else right?
+
+A Ruby process can fork multiple children processes. The memory model for a process is as follows:
 
 ![efficiency_processes_memory_design](/images/efficiency_processes_memory_design.png)
 
-Before we get into that, let's talk a little bit more about processes. Each process has its own address space, stack, and heap. When you fork a process in Ruby, you create a child process which will get its own stack and share the same heap initially. This is called copy-on-write, meaning that the child process will share the same resources with the parent process, until modifications are made to that resource, in which case, that resource will be written into the child's own heap.
+Each process has its own address space, stack, and heap. When a process is forked in Ruby, a child process is created, which will get its own stack. As for the heap, depending on the OS the copy-on-write mechanism is employed:
+- children processes start by referencing the parent's process heap, effectively sharing the same resources with the parent process
+- as modifications to the resources on the heap occur, children processes write the mutated resources to their own heap, leaving the parent's heap unmodified
+
+We benchmarked the memory consumption of having one process and two processes by using a Rails application. Having one process takes up 125MB of memory, but having two processes don't take twice as much memory. This is the copy-on-write mechanism at work.
 
 ![memory_usage_processes](/images/memory_usage_processes.png)
 
-Here, we benchmarked both having one process and two processes. Having one process takes up 125MB of memory, but having two processes don't take twice as much memory. This is the copy-on-write mechanism at work.
+Let's assume the Rails application in our scenario is a pretty big Rails app and takes about 400MB. By forking 4 processes Workerholic, will at most take 1.6GB of memory out of 4GB, meaning, forking 4 processes is a viable option.
 
-As we mentioned previously, using our four CPU cores, we can fork to a total of four processes and reduce our image processing enqueuing:processing ratio down to 1:10. But that is still not good enough. After an extended period of time, we will eventually end up with a backlog and a huge memory footprint. Where do we go from here?
+As we mentioned previously, using our four CPU cores, we can fork to a total of four processes and reduce our image processing enqueuing:processing ratio down to 1:10. But that is still not good enough. After an extended period of time, we will eventually end up with a backlog of jobs, effectively implicating a huge latency and memory footprint. What else can we do to even out or enqueuing:processing ratio?
```

Comment:
As we mentioned previously, using our four CPU cores, we can fork a total of four processes and reduce our image processing enqueuing:processing ratio down to 1:10. But that is still not good enough. After an extended period of time, we will eventually end up with a backlog of jobs, effectively imposing a huge latency and memory footprint. What else can we do to even out or enqueuing:processing ratio?